resources/resource_utils.py

```diff
@@ -0,0 +1,60 @@
+from typing import Optional, Dict
+import hail as hl
+
+
+# Resource classes
+class _Resource:
+    """
+    Generic abstract resource class.
+    """
+    def __init__(self, path: str, source_path: Optional[str] = None, **kwargs):
+        if type(self) is _Resource:
+            raise TypeError("Can't instantiate abstract class _Resource")
+
+        self.source_path = source_path
+        self.path = path
+        self.__dict__.update(kwargs)
+
+    @classmethod
+    def versioned(cls, versions: Dict[str, '_Resource']) -> '_Resource':
+        """
+        Creates a versioned resource.
+        The `path`/`source_path` attributes of the versioned resource are those
+        of default version of the resource (first in the dict keys).
+
+        In addition, all versions of the resource are stored in the `versions` attribute.
+
+        :param dict of str -> _Resource versions: A dict of version name -> resource. The default version should be first in the dict keys.
+        :return: A Resource with versions
+        :rtype: _Resource
+        """
+        latest = versions[list(versions)[0]]
```

Comment:
Using a `default_version` keyword argument would be a clearer way to do this. Dicts don't connote an ordering, so when looking at code where a versioned resource is instantiated, it's not obvious that the order of keys is significant. That would require reading this docstring. Also, it's only in Python 3.7+ that the order of the dictionary keys is guaranteed to be preserved.

Reply:
Yup totally agree!